### ISSUE TITLE
Can the api give a clear warning message if miss config for BatchLogRecordProcessorBuilder's maxQueueSize and maxExportBatchSize?

### ISSUE THREAD

**Issue by tongshushan** [open-telemetry/opentelemetry-java#6454]
Hello,
For **BatchLogRecordProcessorBuilder** configurations， if miss configue **maxQueueSize < maxExportBatchSize**, can the api give a clear warning message? At present it's no hint message, and the logs will be lost.

io.opentelemetry: 1.37.0

related link:
https://github.com/open-telemetry/opentelemetry-java/issues/6443

Thanks.

**Comment by tongshushan** [comment]
Additional: when miss configue **maxQueueSize < maxExportBatchSize**, besides give a warning message, can we set maxExportBatchSize=maxQueueSize , so that can ensure the logs not lost?

**Comment by jkwatson** [comment]
@tongshushan Are you able to put in a PR to address this?

**Comment by chukunx** [comment]
Hey @jkwatson i took a stab at it, please let me know if it looks good  https://github.com/open-telemetry/opentelemetry-java/pull/7045

**Comment by trask** [comment]
I have similar question to @breedx-splk's https://github.com/open-telemetry/opentelemetry-java/issues/7024#issuecomment-2617347391, I'm not clear why maxQueueSize must be greater than maxExportBatchSize to ensure data loss doesn't occur?

at the same time, I agree that I probably would recommend configuring maxQueueSize >= maxExportBatchSize, I'd just like to be clear whether we're recommending this as a _must_ to avoid data loss, or a recommendation / best practice

**Comment by jack-berg** [comment]
These lines of code are the problem: 

- https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L240
- https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L265

Together, they collaborate to mean that the worker thread is never notified that its time to export based on the queue filling up. Instead, it always has to wait for next export time based on `scheduleDelayNanos`. 

And so as the issue poster points out, the seemingly benign mistake of setting `maxExportBatchSize` to be greater than `maxQueueSize` results in data loss as soon the qeueue starts filling up faster than the time allotted in `scheduleDelayNanos`.

We don't necessarily have to throw an exception when the user misconfigures like this, but we need to fix the behavior so that when the queue fills up, the worker is properly signaled to perform an export.

**Comment by trask** [comment]
> Together, they collaborate to mean that the worker thread is never notified that its time to export based on the queue filling up. Instead, it always has to wait for next export time based on `scheduleDelayNanos`.

oh, yikes! I totally missed the wait/notify, I was thinking the queue was continuously drained (but I like the cleverness of limiting the context switching 👍)

**Comment by chukunx** [comment]
Heyy thanks folks for chiming in, I did [some leg work](https://github.com/open-telemetry/opentelemetry-specification/pull/4388#issuecomment-2629539795) to see how other languages are doing on this matter, please let me know if that helps and how you'd like to proceed on this

**Comment by jack-berg** [comment]
Thanks for that @chukunx. I think opentelemetry-java's behavior I described [here](https://github.com/open-telemetry/opentelemetry-java/issues/6454#issuecomment-2620223241) is a bug. Options to fix:

1. Continue to allow `maxExportBatchSize > maxQueueSize`, but fix the bug so that export is triggered when the `maxQueueSize` is reached. Log a warning when `maxExportBatchSize > maxQueueSize`.
2. Throw an exception when `maxExportBatchSize > maxQueueSize`. 

Option 1 represents a more lenient approach. We accept the invalid config and essentially ignore it, since `maxExportBatchSize` doesn't play any roll once `maxQueueSize` is used to trigger export and the queue size drained to the export batch has size `maxQueueSize`.

Option 2 is more rigid, representing the fail fast mentality. 

We generally fail fast in this repo, although this is a bit of a special case because nothing is actually broken (after we fix the bug) when `maxExportBatchSize > maxQueueSize`. In a related situation, we _don't_ throw an exception when users configure an OTLP connection timeout to be greater than the overall request timeout, despite there being no practical value in this config. This suggests option 1 is the solution.

**Comment by chukunx** [comment]
Glad that helped @jack-berg! Option 1 sounds better from backward compatibility point of view as well. 

For implementation I can see two approaches:

a. Add an additional check to this condition 

https://github.com/open-telemetry/opentelemetry-java/blob/cb64451c72a96a65d9d32a997852b87c150fcf2f/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L257

to be something like 

```java
if (batch.size() >= maxExportBatchSize || batch.size() >= maxQueueSize || System.nanoTime() >= nextExportTime)
```

which effectively means `batch.size()` is checked against `min(maxExportBatchSize, maxQueueSize)`.

b. Comparing the two values at the builder and adjusting their values so that the max batch size does not exceeds the max queue size when creating the processor. The [Go implementation](https://github.com/open-telemetry/opentelemetry-go/blob/987efa649e4042bac326d9988b9e08bc770b2d53/sdk/trace/batch_span_processor.go#L88-L94) is a good one to borrow in my opinion. 

```go
	if maxExportBatchSize > maxQueueSize {
		if DefaultMaxExportBatchSize > maxQueueSize {
			maxExportBatchSize = maxQueueSize
		} else {
			maxExportBatchSize = DefaultMaxExportBatchSize
		}
	}
```

Do you have preference over the two approaches?

**Comment by jack-berg** [comment]
> (Second thought on this: data loss can still happen when queue is filling up fast)

Yes data loss can happen with the batch processor. This is necessary to protect an application from unbounded resource utilization. Users can detect data loss and reconfigure (turn off instrumentation, reduce sampling rate, increase batch processor queue size) by looking at the `processSpans` counter where `dropped=true`, which is incremented [here](https://github.com/open-telemetry/opentelemetry-java/blob/cb64451c72a96a65d9d32a997852b87c150fcf2f/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L238). 

> a. Add an additional check to this condition

I would also want to double check that the [batch](https://github.com/open-telemetry/opentelemetry-java/blob/cb64451c72a96a65d9d32a997852b87c150fcf2f/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L233) and [spansNeeded](https://github.com/open-telemetry/opentelemetry-java/blob/cb64451c72a96a65d9d32a997852b87c150fcf2f/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java#L182) fields are being sized / set appropriately, since they are involved in signalling as well.

> Comparing the two values at the builder and adjusting their values so that the max batch size does not exceeds the max queue size when creating the processor.

This is easier to reason about and implement IMO. I think there is a minor semantic difference between the two approaches. The first allows a situation where the worker thread is triggered by the max queue size being reached, but a export batch size bigger than the max queue size since spans may continue to flow and be added to the queue as it is being drained. But I think we should probably ignore this edge case and opt for this simpler solution unless needed.

**Comment by chukunx** [comment]
Hey @jack-berg I've updated https://github.com/open-telemetry/opentelemetry-java/pull/7045 with the option b. let me know if it looks good. 

For a. I was thinking it might be an overkill to introduce changes involving `batch` and `spansNeeded` (to be matching with the maxQueueSize when it's smaller than the maxExportBatchSize) so that the processor explicitly handles this scenario. It might end up looking confusing as well for folks who aren't familiar with such edge case. However, if there's an emerging use case for it, I can take a stab in a separate PR.
